Update build.md (#1221)
3scale/APIcast: doc/policies.md

# APIcast policies

The behavior of APIcast is customizable via policies. A policy basically tells

> APIcast what it should do in each of the nginx phases. This document details

Reviewer: ```suggestion
APIcast what it should do in each of the NGINX phases. This document details
```

describes what to do in the `rewrite` and `header_filter` phases and a policy B
that describes what to run in `access` and `header_filter`. Assume also that
when describing the chain, we indicate that policy A should be run before
policy B. 
When APIcast receives an HTTP request, it will check the policy chain

> perform the tasks outlined in the following table:

Reviewer: ```suggestion
and perform the tasks outlined in the following table:
```

described to see what it should run on each phase:
- rewrite: execute the function policy A provides for this phase.
- access: execute the function policy B provides for this phase.

> - content: do nothing. Neither policy A nor B describes what to do.

Reviewer: ```suggestion
- content: Does nothing. Neither policy A nor B describes what to do.
```

I'm not sure if you want to indicate that the user shouldn't do anything, or the phase does nothing. Can you please reword this and the next line?

Author: It refers to the phase as per the context.

Reviewer: To facilitate reading, rephrase the last sentence and add a table to describe what APIcast does. This is an example:

```
When APIcast receives an HTTP request, it will check the policy chain
perform the tasks outlined in the following table:

| Phase    | APIcast tasks                                           |
|----------|---------------------------------------------------------|
| rewrite  | Executes the function policy A provides for this phase. |
| access   | Executes the function policy B provides for this phase. |
| content  | None. Neither policy A nor B  describes what to do.     |
| balancer | None. Neither policy A nor B  describes what to do.     |
```

Author: The tabular representation does make it look better and more understandable. Made the changes. Please check.

- balancer: do nothing. Neither policy A nor B describes what to do.
- header_filter: execute first the function policy A provides for this phase
  and then the function policy B provides for this phase. Remember that policy
  chains define an order, and we specified that policy A comes before policy B.
- body_filter: do nothing. Neither policy A nor B describes what to do.
- post_action: do nothing. Neither policy A nor B describes what to do.
- log: do nothing. Neither policy A nor B describes what to do.

> Notice that we did not indicate what APIcast does in the `init` and the

Reviewer: ```suggestion
Notice that there is not a description of what APIcast does in the `init` and the
```

### Policy structure

The policy is expected to have some structure, so APIcast can find it. Minimal policy structure consists of two files: `init.lua` and `apicast-policy.json`.

Custom policies are expected to be on the following paths:

* `APICAST_DIR/policies/${name}/${version}/`

And builtin ones also on:

* `APICAST_DIR/src/apicast/policy/${name}/`

> All files in the policy directory are namespaced, so you can vendor dependencies.  Consider the following structure:

Reviewer: ```suggestion
All files in the policy directory are namespaced, so you can move the dependencies to the `/vendor' directory.  Consider the following structure:
```

And actually implement everything in `my_stuff.lua`. This makes the policy file easier to find by humans.

> Let's say the policy needs some 3rd party dependency. Those can be put anywhere in the policy structure and will be available. Try to imagine it as UNIX `chroot`. So for example `require('vendor/dependency')` will load `APICAST_DIR/policies/my_stuff/1.0/vendor/dependency.lua` when loading your policy.

Reviewer: ```suggestion
If the policy requires third-party dependencies. Those can be put anywhere in the policy structure and will be available. Imagine it as UNIX `chroot`. So for example `require('vendor/dependency')` will load `APICAST_DIR/policies/my_stuff/1.0/vendor/dependency.lua` when loading your policy.
```
